File: fc-nodejs/benchmark_fcdefinition.js
import {function_call_playground_definition} from "./FCdefinition.js";
import {BENCHMARK_ITERATIONS, BENCHMARK_RESULTS_DIR} from "./config.js";
import path from 'path';
import fs from 'fs';

const prompt = [
    "what is the definition of 'hello'?",
    "what is the definition of 'world'?",
    "what is the definition of 'catastrophe'?",
    "what is the definition of 'exectution'?",
]

const iterations = BENCHMARK_ITERATIONS;

const run_benchmark = async()=>{
    console.log("Running benchmark for FCdefinition...");
    const results = [];
    for(const p of prompt){
        for (let i=0; i<iterations; i++){
            console.log("Prompt:", p);
            const start = Date.now();
            const result = await function_call_playground_definition(p);
            const end = Date.now();
            const timeTaken = (end-start)/1000;
            results.push({
                prompt:p,
                timeTaken:timeTaken,
                result:result
            })
        }
    }
    return results;
}

const success_rate = (results)=>{
    const success_count = results.filter(r=>r.result.isValid).length;
    return (success_count/results.length)*100;
}

const min_time = (results)=>{
    return Math.min(...results.map(r=>r.timeTaken));
}

const max_time = (results)=>{
    return Math.max(...results.map(r=>r.timeTaken));
}

const avg_time = (results)=>{
    return results.reduce((sum,r)=>sum+r.timeTaken,0)/results.length;
}

const median_time = (results)=>{
    const sorted = results.map(r=>r.timeTaken).sort((a,b)=>a-b);
    const mid = Math.floor(sorted.length/2);
    return sorted[mid];
}

const std_dev = (results)=>{
    const avg = avg_time(results);
    const variance = results.reduce((sum,r)=>sum+(r.timeTaken-avg)**2,0)/results.length;
    return Math.sqrt(variance);
}

const save_results = (results)=>{
    const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
    const filename = `fcdefinition_benchmark_${timestamp}.json`;
    
    // Create the directory if it doesn't exist
    if (!fs.existsSync(BENCHMARK_RESULTS_DIR)) {
        fs.mkdirSync(BENCHMARK_RESULTS_DIR, { recursive: true });
    }
    
    const filepath = path.join(BENCHMARK_RESULTS_DIR, filename);
    fs.writeFileSync(filepath, JSON.stringify(results, null, 2));
    console.log(`Results saved to: ${filepath}`);
}

export const main_definition = async()=>{
    const results = await run_benchmark();
    const successRate = success_rate(results);
    const minTime = min_time(results);
    const maxTime = max_time(results);
    const avgTime = avg_time(results);
    const medianTime = median_time(results);
    const stdDev = std_dev(results);
    console.log(`Benchmark Results:`);
    console.log(`Success Rate: ${successRate.toFixed(2)}%`);
    console.log(`Min Time: ${minTime.toFixed(2)}s`);
    console.log(`Max Time: ${maxTime.toFixed(2)}s`);
    console.log(`Avg Time: ${avgTime.toFixed(2)}s`);
    console.log(`Median Time: ${medianTime.toFixed(2)}s`);
    console.log(`Std Dev: ${stdDev.toFixed(2)}s`);
    save_results(results);
}

main_definition();

File: function-calling/fc-python/benchmark_fcdefinition.py
import time
import json
import statistics
from datetime import datetime
import sys
import os
from config import BENCHMARK_ITERATIONS

# Import the function_call_playground directly since we're in the same directory
from FCdefinition import function_call_playground

# Create benchmark_results directory if it doesn't exist
results_dir = os.path.join(os.path.dirname(os.path.abspath(__file__)), "benchmark_results")
os.makedirs(results_dir, exist_ok=True)

test_cases = {
    "what is the definition of the word 'good'": {
        "expected_output": "The definition of good is a positive or favorable opinion or judgment."
    },
    "what is the definition of the word 'bad'": {
        "expected_output": "The definition of bad is a negative or unfavorable opinion or judgment."
    },
    "what is the definition of the word 'eviscerate'": {
        "expected_output": "The definition of eviscerate is to remove the internal organs of (a living animal) by cutting through the body cavity."
    },
    "what is the definition of the word 'hashtable'": {
        "expected_output": "A hashtable is a data structure that stores key-value pairs in an array."
    },
    "what is the definition of the word 'yeet'": {
        "expected_output": "The definition of yeet is to throw something or someone away."
    },
    "what is the definition of the word 'bet'": {
        "expected_output": "The definition of bet is to place a wager on something."
    },
    "what is the definition of the word 'asshole'": {
        "expected_output": "The definition of asshole is a derogatory term for a person who is considered to be unpleasant."
    }
}

results = {
    "timestamp": datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
    "test_cases": [],
    "summary": {}
}

for test_case in test_cases:
    print(f"Running benchmark for {test_case}...")
    
    execution_times = []
    success_count = 0
    
    for i in range(BENCHMARK_ITERATIONS):
        try:
            start_time = time.time()
            response = function_call_playground(test_case)
            end_time = time.time()
            
            execution_time = end_time - start_time
            execution_times.append(execution_time)
            success_count += 1
            
            print(f"Iteration {i+1}/{BENCHMARK_ITERATIONS}: {execution_time:.2f} seconds")
        except Exception as e:
            print(f"Iteration {i+1}/{BENCHMARK_ITERATIONS} failed: {str(e)}")
    
    # Calculate statistics for this test case
    if execution_times:
        test_result = {
            "name": test_case,
            "success_rate": success_count / BENCHMARK_ITERATIONS * 100,
            "max_time": max(execution_times),
            "min_time": min(execution_times),
            "avg_time": statistics.mean(execution_times),
            "median_time": statistics.median(execution_times),
            "std_dev": statistics.stdev(execution_times) if len(execution_times) > 1 else 0
        }
    else:
        test_result = {
            "name": test_case,
            "success_rate": 0,
            "max_time": None,
            "min_time": None,
            "avg_time": None,
            "median_time": None,
            "std_dev": None
        }
    
    results["test_cases"].append(test_result)

# Calculate overall summary
all_times = [t["avg_time"] for t in results["test_cases"] if t["avg_time"] is not None]

if all_times:
    results["summary"] = {
        "overall_avg_time": statistics.mean(all_times),
        "overall_min_time": min(all_times),
        "overall_max_time": max(all_times),
        "overall_median_time": statistics.median(all_times),
        "overall_std_dev": statistics.stdev(all_times) if len(all_times) > 1 else 0
    }

# Save results to file in the benchmark_results directory
results_file = os.path.join(results_dir, f"benchmark_results_{datetime.now().strftime('%Y%m%d_%H%M%S')}.json")
with open(results_file, "w") as f:
    json.dump(results, f, indent=4)

print(f"\nResults saved to: {results_file}")

# Print summary
print("\n=== BENCHMARK SUMMARY ===")
print(f"Timestamp: {results['timestamp']}")
print("\nTest Case Results:")
for test in results["test_cases"]:
    print(f"\n{test['name']}:")
    print(f"  Success Rate: {test['success_rate']:.1f}%")
    if test['avg_time'] is not None:
        print(f"  Average Time: {test['avg_time']:.2f} seconds")
        print(f"  Min Time: {test['min_time']:.2f} seconds")
        print(f"  Max Time: {test['max_time']:.2f} seconds")
        print(f"  Median Time: {test['median_time']:.2f} seconds")
        print(f"  Standard Deviation: {test['std_dev']:.2f} seconds")
    else:
        print("  All iterations failed")

if results["summary"]:
    print("\nOverall Summary:")
    print(f"  Overall Average Time: {results['summary']['overall_avg_time']:.2f} seconds")
    print(f"  Overall Min Time: {results['summary']['overall_min_time']:.2f} seconds")
    print(f"  Overall Max Time: {results['summary']['overall_max_time']:.2f} seconds")
    print(f"  Overall Median Time: {results['summary']['overall_median_time']:.2f} seconds")
    print(f"  Overall Standard Deviation: {results['summary']['overall_std_dev']:.2f} seconds")

 

